keps/sig-node/20190425-node-readiness-gates.md

```diff
@@ -0,0 +1,360 @@
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+As of Kubernetes v1.12, the health of extended node components like storage, network, device plugins, logging
+or monitoring agents aren't taken into account while managing user pods on kubernetes nodes. This leads to a
+broken contract between cluster administrators and end users where the latter ends up debugging issues like
+lost metrics or logs, lack of network or storage connectivity, etc. On top of that lack of a generic abstraction
+for node readiness leads to hacks across the k8s ecosystem to deal with the asynchronous & dynamic nature
+of bootsrapping and managing k8s nodes.
+
+This proposal presents several approaches for solving the problem of extended node health or node readiness gates.
+
+## Motivation
+
+Since the early days of kubernetes, kubelet has been the primary source of readiness of a node where it queried
+the health of critical system components like the docker daemon, network components, etc. Kubelet owns the node “Ready”
+condition and the rest of Kubernetes cluster components rely on this (and a few other conditions) to infer that a node is “online”.
+
+Kubernetes has evolved quite a bit since the early days and now a kubernetes “node” is comprised of many
+system microservices that are all essential for a node to go “online” and start running user applications.
+Examples include device & storage plugins (network plugins in the future?), logging & monitoring agents,
+networking daemons like kube-proxy, machine management daemons like Node Problem Detector, etc. Over the years,
+kubelet and the container runtimes have evolved adequately to become highly reliable and so they become “ready”
+more consistently. But the other newer system extensions may not be as mature yet, or may be slow to transition
+to “ready” for legitimate reasons.
+
+The lack of a single unambiguous signal on when a “node” is “ready”, beyond just the kubelet or a container runtime
+is causing several issues today.
+
+* System components like Cluster Autoscaler and Scheduler have to combine several Node Conditions to evaluate a node’s
```

Comment:
Can we add the use-case in https://github.com/kubernetes/kubernetes/pull/33573 ?

e.g.

* External systems (such as networking) must often be configured before pods should be scheduled on a node.